### ISSUE TITLE
GrabDragInteraction doesn't work with VoiceOver

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/scenery-phet#688]
Initially discovered in https://github.com/phetsims/balloons-and-static-electricity/issues/519.

Once the object has been picked up and released, returning to it in a grabbable state causes VoiceOver to fail to read the accessible name and role information. Instead it just says "application".

This was observed in the balloons in BASE but also everything using the GrabDragInteraction in macOS VoiceOver in Safari. We confirmed that this is not happening in macOS Firefox, so this is a Safari bug.

This is not happening in master in High Sierra 10.13.6. It is also not happening in the latest version of Safari in the published version of BASE.

In https://github.com/phetsims/balloons-and-static-electricity/issues/519, it was recommended that we try
- Adding an aria-label as a workaround
- See if the implementation of the grab/drag interaction as it was last time BASE was published works. Before this, see if a vanilla HTML example can be used to test behavior before changing common code.

This blocks publication of sims with Interactive Description that use GrabDragInteraction.

**Comment by jessegreenberg** [comment]
![image](https://user-images.githubusercontent.com/6396244/128382216-083ea409-2bad-4a0b-98d7-9ceb0602f4b5.png)

Check out the above image! The application role is still on the button according to the dev tools, though it is not in the markup.

**Comment by jessegreenberg** [comment]
Oddly, the bug where the "Application" role persists in the dev tools only happens when VoiceOver is enabled.

**Comment by jessegreenberg** [comment]
I tried explicitly setting the role to something else (rather than just removing `application`) and it still kept "application" according to the dev tools when VoiceOver is enabled.

**Comment by jessegreenberg** [comment]
The highlight that appears from VoiceOver is where the inner text for the Grabbable button is positioned
![image](https://user-images.githubusercontent.com/6396244/128415273-7ed87e3c-dd62-4153-88a8-ba9efbb7222e.png)

Oddly, when focus is on the balloon in this case the hint I get is "You are on a text element" and the text element I am on is the inner text of the "Grab Yellow Balloon" button. This is making me wonder if it is related to styling.

**Comment by jessegreenberg** [comment]
I made this HTML example to try to find out what was causing the problem. But even as I added more and more attributes and made the example more complicated VoiceOver always reported the name and role of the "Grab" button. So this is indicating that it is not caused by the switch to use one element rather than move focus between two DOM elements.
<details>

```html
<html>
<body>

<article id="parent" style="position: absolute;">
  <button id="grab-drag-element">Grab Me</button>
  <button>Something else</button>
</article>

</body>

<script>

  let isGrabbable = true;
  const parent = document.getElementById( "parent" );
  const swapRepresentation = () => {

    const oldElement = document.getElementById( 'grab-drag-element' );
    parent.removeChild( oldElement );

    let newElement;
    if ( isGrabbable ) {

      // we are turning to draggable
      newElement = document.createElement( 'div' );
      newElement.setAttribute( 'role', 'application' );
      newElement.id = 'grab-drag-element';
      newElement.tabIndex = 0;
      newElement.innerText = "Drag me!";
      newElement.setAttribute( 'aria-roledescription', "Draggable" );
      newElement.setAttribute( 'aria-label', "Drag me!" );

      // hack alert - wait a second to add this listener so that when focus changes
      // immediately we don't get a keyup event right away
      // window.setTimeout( () => {
        newElement.addEventListener( 'keyup', event => {
          if ( event.key === 'Enter' ) {
            swapRepresentation();
          }
        } );
      // }, 1000 );
    }
    else {

      // we are turning to grabbable
      newElement = document.createElement( 'button' );
      newElement.textContent = 'Grab Me';

      newElement.addEventListener( 'click', () => {
        console.log( 'click' );
        swapRepresentation();
      } );
    }

    newElement.setAttribute( 'style', "position: fixed; top: 474.138px; left: 300.09px; width: 80.9583px; height: 134.729px;" );
    newElement.id = 'grab-drag-element';
    parent.appendChild( newElement );
    newElement.focus();

    isGrabbable = !isGrabbable;
  }


  // add listener to the initial element
  document.getElementById( 'grab-drag-element' ).addEventListener( 'click', swapRepresentation );

</script>
</html>
```
</details>

**Comment by jessegreenberg** [comment]
Something about this patch has the bug fixed:
<details>

```patch
Index: balloons-and-static-electricity/js/balloons-and-static-electricity/view/BalloonNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- balloons-and-static-electricity/js/balloons-and-static-electricity/view/BalloonNode.js	(revision 94fb4106581d99cd4500358f4073335e3a20a7b4)
+++ balloons-and-static-electricity/js/balloons-and-static-electricity/view/BalloonNode.js	(date 1628182852402)
@@ -365,6 +365,10 @@
         centerTop: balloonImageNode.centerBottom.plusXY( 0, 10 )
       },
 
+      grabbableOptions: {
+        // ariaLabel: 'Test Label'
+      },
+
       keyboardHelpText: grabBalloonKeyboardHelpString,
 
       onGrab: () => {
Index: scenery-phet/js/accessibility/GrabDragInteraction.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- scenery-phet/js/accessibility/GrabDragInteraction.js	(revision ad556223e733176e4496c14fa91409f4d2ca4bfe)
+++ scenery-phet/js/accessibility/GrabDragInteraction.js	(date 1628196475330)
@@ -212,7 +212,7 @@
       ariaRole: 'application',
 
       // to cancel out grabbable
-      containerTagName: null
+      // containerTagName: null
     }, options.draggableOptions );
 
     // @private
@@ -224,15 +224,15 @@
     assert && assert( !options.grabbableOptions.innerContent, 'GrabDragInteraction sets its own innerContent, see objectToGrabString' );
 
     options.grabbableOptions = merge( {
-      containerTagName: 'div',
+      // containerTagName: 'div',
       ariaRole: null,
       tagName: 'button',
 
       // position the PDOM elements when grabbable for drag and drop on touch-based screen readers
       positionInPDOM: true,
 
-      accessibleName: null,
-      ariaLabel: null // also since many use ariaLabel to set accessibleName
+      //accessibleName: null,
+      //ariaLabel: null // also since many use ariaLabel to set accessibleName
     }, options.grabbableOptions );
 
     // @private
@@ -498,24 +498,24 @@
     this.grabbable = true;
 
     // To support gesture and mobile screen readers, we change the roledescription, see https://github.com/phetsims/scenery-phet/issues/536
-    if ( this.supportsGestureDescription ) {
-      this.node.setPDOMAttribute( 'aria-roledescription', movableString );
-    }
-    else if ( this.node.hasPDOMAttribute( 'aria-roledescription' ) ) {
-
-      // by default, the grabbable has no roledescription. Can be overwritten in `onGrabbable()`
-      this.node.removePDOMAttribute( 'aria-roledescription' );
-    }
+    // if ( this.supportsGestureDescription ) {
+    //   this.node.setPDOMAttribute( 'aria-roledescription', movableString );
+    // }
+    // else if ( this.node.hasPDOMAttribute( 'aria-roledescription' ) ) {
+    //
+    //   // by default, the grabbable has no roledescription. Can be overwritten in `onGrabbable()`
+    //   this.node.removePDOMAttribute( 'aria-roledescription' );
+    // }
 
-    if ( this.addAriaDescribedbyPredicate( this.numberOfGrabs ) ) {
-
-      // this node is aria-describedby its own description content, so that the description is read automatically
-      // when found by the user
-      !this.node.hasAriaDescribedbyAssociation( this.descriptionAssociationObject ) && this.node.addAriaDescribedbyAssociation( this.descriptionAssociationObject );
-    }
-    else if ( this.node.hasAriaDescribedbyAssociation( this.descriptionAssociationObject ) ) {
-      this.node.removeAriaDescribedbyAssociation( this.descriptionAssociationObject );
-    }
+    // if ( this.addAriaDescribedbyPredicate( this.numberOfGrabs ) ) {
+    //
+    //   // this node is aria-describedby its own description content, so that the description is read automatically
+    //   // when found by the user
+    //   !this.node.hasAriaDescribedbyAssociation( this.descriptionAssociationObject ) && this.node.addAriaDescribedbyAssociation( this.descriptionAssociationObject );
+    // }
+    // else if ( this.node.hasAriaDescribedbyAssociation( this.descriptionAssociationObject ) ) {
+    //   this.node.removeAriaDescribedbyAssociation( this.descriptionAssociationObject );
+    // }
 
     this.baseInteractionUpdate( this.grabbableOptions, this.listenersForDragState, this.listenersForGrabState );
 
@@ -534,7 +534,7 @@
     this.grabbable = false;
 
     // by default, the draggable has roledescription of "movable". Can be overwritten in `onDraggable()`
-    this.node.setPDOMAttribute( 'aria-roledescription', movableString );
+    //this.node.setPDOMAttribute( 'aria-roledescription', movableString );
 
     // This node is aria-describedby its own description content only when grabbable, so that the description is
     // read automatically when found by the user with the virtual cursor. Remove it for draggable
```

</details>

**Comment by jessegreenberg** [comment]
I identified that this line is causing this bug, removing this line makes the issue go away:
https://github.com/phetsims/scenery-phet/blob/4e822fb1d08ba5a325196abdb5886a0af5a793a3/js/accessibility/GrabDragInteraction.js#L235

**Comment by jessegreenberg** [comment]
Setting the aria-label in BASE **does** make this issue go away. I don't know why I found that it didnt' earlier, I probably forgot to clear caches in that test. The button in my test HTML also does not have an aria-label so I don't know why it is working there.

So I think we need to make sure that aria-label is set in the `grabbableOptions` as a workaround for this bug.

**Comment by jessegreenberg** [comment]
I am confused again.

With this line
```js
options.grabbableOptions.ariaLabel = this.grabbableAccessibleName;
```

VoiceOver calls the balloon "Grab Yellow Balloon, application"

Without that line and with ariaLabel removed from options, it says "Yellow Balloon, button". It is using the aria-label from the previous element, in this case. For some reason changing the aria-label seems to be causing problems.

**Comment by jessegreenberg** [comment]
I can get VoiceOver to say "Grab Yellow Balloon, button" by assigning an `aria-roledescription` of "button" in its grabbable state.

There is some weird VoiceOver bugginess causing all of this and I don't think it is worth our time to investigate more. But I will summarize the combinations of attributes I tried with their output.

The first time VO reads the grabbable, it sounds correct. The following markup and output combinations are after pickup and release.
```<button>Grab Yellow Balloon</button>``` -> "Web application"
```<button aria-label="Grab Yellow Balloon">Grab Yellow Balloon</button>``` -> "Grab Yellow Balloon, application"
```<button aria-label="Grab Yellow Balloon" role="button">Grab Yellow Balloon</button>``` -> "Grab Yellow Balloon, application"
```<button aria-label="Grab Yellow Balloon" aria-roledescription="button">Grab Yellow Balloon</button>``` -> "Grab Yellow Balloon, button"

**Comment by jessegreenberg** [comment]
I tested this change with NVDA + Firefox, JAWS + Chrome and Safari + VoiceOver, each are working and sounding good.

@zepumph would you mind reviewing https://github.com/phetsims/scenery-phet/commit/ff02afa159b11320e841e73a9cc56b1e18b3bcf8 before closing this issue?

**Comment by zepumph** [comment]
This does not feel like a robust solution. I'd like to discuss alternatives, potentially in person. Basically you are mixing higher-level API features with lower-level ones. In master this works well because of the current default accessibleName behavior, but it doesn't feel too maintainable to me. Instead, we should expand the assertions about how what lower level keys you can pass in (to include `innerContent` and perhaps more) (by noting the following . . . )

https://github.com/phetsims/scenery/blob/cacb390e636c006c8947d9d07cc6a52d8f0540b9/js/accessibility/pdom/ParallelDOM.js#L2932-L2949


Then we can supply our own accessibleName behavior that always sets aria-label (asserting that we clients don't pass that in either). And maybe innerContent? This is what the the higher level API was designed to support. Thoughts?

**Comment by jessegreenberg** [comment]
Sorry, I don't understand, are you recommending a general improvement for GrabDragInteraction? I actually don't see higher level API funtions used in GrabDragInteraction yet, other than clearing the `accessibleName` in `grabbableOptions`. Or should `accessibleName` be removed from `grabbableOptions`, I don't see `accessibleName` set in `draggableOptions` or passed to grabbableOptions in the project.

**Comment by jessegreenberg** [comment]
@zepumph and I discussed this issue and determined that it is OK as is. There isn't a mix of high and low level API functions, but there was some confusion because `draggableAccessibleName` is not actually using `accessibleName` setters.

On the call we agreed this could be closed.
